framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,670 @@
+            return new Range(Collections.min(possibleValues), Collections.max(possibleValues));
+        } else {
+            List<BigInteger> bigPossibleValues =
+                    Arrays.asList(
+                            BigInteger.valueOf(from).multiply(BigInteger.valueOf(right.from)),
+                            BigInteger.valueOf(from).multiply(BigInteger.valueOf(right.to)),
+                            BigInteger.valueOf(to).multiply(BigInteger.valueOf(right.from)),
+                            BigInteger.valueOf(to).multiply(BigInteger.valueOf(right.to)));
+            BigInteger bigFrom = Collections.min(bigPossibleValues);
+            BigInteger bigTo = Collections.max(bigPossibleValues);
+            return bigRange2LongRange(bigFrom, bigTo);
+        }
+    }
+
+    /**
+     * Returns the smallest range that includes all possible values resulting from dividing (integer
+     * division) an arbitrary value in this range by an arbitrary value in the specified range. We
+     * call this the division of two ranges.
+     *
+     * @param right the specified range by which this range is divided
+     * @return the range resulting from dividing this range by the specified range
+     */
+    public Range divide(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        if (right.contains(0)) {
+            // Should warn division by zero
+            return EVERYTHING;
+        }
+
+        // Special cases that involve overflow.
+        // The only overflow in integer division is Long.MIN_VALUE / -1 == Long.MIN_VALUE.
+        if (from == Long.MIN_VALUE && right.contains(-1)) {
+            // The values in the right range are all negative because right does not contain 0 but
+            // does contain 1.
+            if (from != to) {
+                // Special case 1:
+                // This range contains Long.MIN_VALUE and Long.MIN_VALUE + 1, which makes the
+                // result range EVERYTHING.
+                return EVERYTHING;
+            } else if (right.from != right.to) {
+                // Special case 2:
+                // This range contains only Long.MIN_VALUE, and the right range contains at least -1
+                // and -2. The result range is from Long.MIN_VALUE to Long.MIN_VALUE / -2.
+                return new Range(Long.MIN_VALUE, Long.MIN_VALUE / -2);
+            } else {
+                // Special case 3:
+                // This range contains only Long.MIN_VALUE, and right contains only -1.
+                return new Range(Long.MIN_VALUE, Long.MIN_VALUE);
+            }
+        }
+
+        long resultFrom;
+        long resultTo;
+        // We needn't worry about the overflow issue starting from here.
+        // To facilitate the calculation of the result range, we categorize all the scenarios into 9
+        // different cases:
```

Comment:
Actually, only `Range.time` and `Range.remainder` are using the approach that we compute many possible values and then use their min and max as the bound. Other operations (add, minus, unary ops...) are using a similar approach as in here.

The reason for determining the range by categorising and compute accordingly is that, in this way the computation is very efficient and the result is the most precise -- it can be proved that the result of the `Range.divide` is "the smallest range that includes all possible values resulting from dividing".

As for `Range.times`, using the approach you mentioned (min max bound) is as precise as this approach here without sacrificing the computational efficiency (we only need to calculate 4 possible values). The code is cleaner that way IMHO so that's why I use the "min max bound" approach there. 

As for `Range.remainder`, it is too complicated (for me) to categorise the cases so that I couldn't compute the accordingly. I have to loosen the resulted range by including possible values that are more than necessary to make sure that the resulted range does cover all possible values, but it would not be "the smallest range" anymore (correct, but not precise).